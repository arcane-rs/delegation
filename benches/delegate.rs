use criterion::{criterion_group, criterion_main, Criterion};
use delegation::delegate;

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

impl AsStr for String {
    fn as_str(&self) -> &str {
        self
    }
}

macro_rules! def_enum {
    ($($variant:ident),+) => {
        #[delegate(derive(AsStr))]
        enum Name {
            $($variant(String)),+
        }
    };
}

macro_rules! def_bench {
    ($($variant:ident),*) => {
        fn bench(bencher: &mut ::criterion::Bencher) {
            $(
                bencher.iter(|| {
                    let name = ::criterion::black_box(
                        Name::$variant(String::from("Test"))
                    );

                    assert_eq!(name.as_str(), "Test");
                });
            )*
        }
    };
}

macro_rules! define {
    ($($variant:ident),*) => {
        def_enum!($($variant),*);
        def_bench!($($variant),*);
    };
}

define!(
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y,
    Z
);

fn delegate_benchmark(c: &mut Criterion) {
    c.bench_function("delegate", bench);
}

criterion_group!(benches, delegate_benchmark);
criterion_main!(benches);
