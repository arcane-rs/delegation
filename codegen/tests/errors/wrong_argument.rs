use delegation::delegate;

#[delegate(derive(AsStr))]
trait AsStr {
    fn as_str(&self) -> &str;
}

#[delegate]
trait AsString {
    fn as_str(&self) -> &str;
}

#[delegate(for(AsString))]
struct FirstName(String);

#[delegate(derive(AsString))]
struct MiddleName(#[delegate(derive(AsString))] String);

#[delegate(derive(AsString))]
struct NickName(#[delegate(for(AsString))] String);

#[delegate(derive(AsString))]
enum OlegName {
    #[delegate(derive(AsString))]
    First(String),
}

#[delegate(derive(AsString))]
enum BorisFullName {
    #[delegate(for(AsString))]
    FirstAndLast(String),
}

#[delegate(derive(AsString))]
enum Name {
    First(#[delegate(derive(AsString))] String),
}

#[delegate(derive(AsString))]
enum FullName {
    FirstAndLast(#[delegate(for(AsString))] String),
}

#[delegate]
enum AliasName {
    Alias(#[delegate(as = SomeType)] String),
}

#[delegate(as = SomeType)]
enum UserName {
    Name(String),
}

#[delegate(as = SomeType)]
struct MemberName(String);

fn main() {
    unreachable!()
}
