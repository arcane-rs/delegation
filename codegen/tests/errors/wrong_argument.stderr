error: unexpected attribute argument
 --> tests/errors/wrong_argument.rs:3:12
  |
3 | #[delegate(derive(AsStr))]
  |            ^^^^^^

error: unexpected attribute argument
  --> tests/errors/wrong_argument.rs:13:12
   |
13 | #[delegate(for(AsString))]
   |            ^^^

error: unexpected attribute argument
  --> tests/errors/wrong_argument.rs:17:30
   |
17 | struct MiddleName(#[delegate(derive(AsString))] String);
   |                              ^^^^^^

error: unexpected attribute argument
  --> tests/errors/wrong_argument.rs:20:28
   |
20 | struct NickName(#[delegate(for(AsString))] String);
   |                            ^^^

error: unexpected attribute argument
  --> tests/errors/wrong_argument.rs:24:16
   |
24 |     #[delegate(derive(AsString))]
   |                ^^^^^^

error: unexpected attribute argument
  --> tests/errors/wrong_argument.rs:30:16
   |
30 |     #[delegate(for(AsString))]
   |                ^^^

error: expected non-macro attribute, found attribute macro `delegate`
  --> tests/errors/wrong_argument.rs:36:13
   |
36 |     First(#[delegate(derive(AsString))] String),
   |             ^^^^^^^^ not a non-macro attribute

error: expected non-macro attribute, found attribute macro `delegate`
  --> tests/errors/wrong_argument.rs:41:20
   |
41 |     FirstAndLast(#[delegate(for(AsString))] String),
   |                    ^^^^^^^^ not a non-macro attribute

error: expected non-macro attribute, found attribute macro `delegate`
  --> tests/errors/wrong_argument.rs:46:13
   |
46 |     Alias(#[delegate(as = "SomeType")] String),
   |             ^^^^^^^^ not a non-macro attribute

error: unexpected attribute argument
  --> tests/errors/wrong_argument.rs:49:12
   |
49 | #[delegate(as = "SomeType")]
   |            ^^

error: unexpected attribute argument
  --> tests/errors/wrong_argument.rs:54:12
   |
54 | #[delegate(as = "SomeType")]
   |            ^^

error[E0277]: the trait bound `String: AsString` is not satisfied
  --> tests/errors/wrong_argument.rs:34:1
   |
34 | #[delegate(derive(AsString))]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `AsString` is not implemented for `String`
   |
   = help: the following other types implement trait `AsString`:
             FullName
             Name
             delegation::__macros::Wrapper<__Delegate>
note: required for `Either<String, Void>` to implement `__delegate_AsString__DelegateOwned`
  --> tests/errors/wrong_argument.rs:8:1
   |
8  | #[delegate]
   | ^^^^^^^^^^^
9  | trait AsString {
   |       ^^^^^^^^ unsatisfied trait bound introduced here
note: required for `delegation::__macros::Wrapper<Name>` to implement `AsString`
  --> tests/errors/wrong_argument.rs:8:1
   |
8  | #[delegate]
   | ^^^^^^^^^^^
9  | trait AsString {
   |       ^^^^^^^^ unsatisfied trait bound introduced here
   = note: this error originates in the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `String: AsString` is not satisfied
  --> tests/errors/wrong_argument.rs:39:1
   |
39 | #[delegate(derive(AsString))]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `AsString` is not implemented for `String`
   |
   = help: the following other types implement trait `AsString`:
             FullName
             Name
             delegation::__macros::Wrapper<__Delegate>
note: required for `Either<String, Void>` to implement `__delegate_AsString__DelegateOwned`
  --> tests/errors/wrong_argument.rs:8:1
   |
8  | #[delegate]
   | ^^^^^^^^^^^
9  | trait AsString {
   |       ^^^^^^^^ unsatisfied trait bound introduced here
note: required for `delegation::__macros::Wrapper<FullName>` to implement `AsString`
  --> tests/errors/wrong_argument.rs:8:1
   |
8  | #[delegate]
   | ^^^^^^^^^^^
9  | trait AsString {
   |       ^^^^^^^^ unsatisfied trait bound introduced here
   = note: this error originates in the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
